lib/pow_assent/ecto/user_identities/context.ex

```diff
@@ -113,18 +124,31 @@ defmodule PowAssent.Ecto.UserIdentities.Context do
     |> repo(config).one()
   end
 
+  # TODO: Remove by 0.4.0
+  @doc false
+  @deprecated "Use `upsert/3` instead"
+  @spec create(user(), user_identity_params(), Config.t()) :: {:ok, user_identity()} | {:error, {:bound_to_different_user, changeset()}} | {:error, changeset()}
+  def create(user, user_identity_params, config), do: upsert(user, user_identity_params, config)
+
   @doc """
-  Creates a user identity.
+  Upserts a user identity.
 
-  User schema module and repo module will be fetched from config.
+  If a matching user identity already exists for the user, the identity will be updated,
+  otherwise a new identity is inserted.
+
+  Repo module will be fetched from config.
   """
-  @spec create(user(), user_identity_params(), Config.t()) :: {:ok, user_identity()} | {:error, {:bound_to_different_user, changeset()}} | {:error, changeset()}
-  def create(user, user_identity_params, config) do
-    user_identity = Ecto.build_assoc(user, :user_identities)
+  @spec upsert(user(), user_identity_params(), Config.t()) :: {:ok, user_identity()} | {:error, {:bound_to_different_user, changeset()}} | {:error, changeset()}
+  def upsert(user, user_identity_params, config) do
+    params                                   = to_string_params(user_identity_params)
+    {uid_provider_params, additional_params} = Map.split(params, ["uid", "provider"])
 
-    user_identity
-    |> user_identity.__struct__.changeset(user_identity_params)
-    |> Context.do_insert(config)
+    user
+    |> get_for_user(uid_provider_params, config)
+    |> case do
+      nil      -> insert_identity(user, params, config)
+      identity -> update_identity(identity, additional_params, config)
```

Comment:
Is there a reason you're not using the database upsert functionality here?

Reply:
https://hexdocs.pm/ecto/Ecto.Repo.html#c:insert/2-upserts

Reply:
I initially did that with your proposal in #69, but I couldn't get it to work. So I decided to just get it to work in a way that was easy for me to understand, and then eventually refactor it to let the DB handle it. I'll try it again now that I've cleaned up the other logic, and ensured that the tests are working. It would be great with ecto upsert 😄

Reply:
Just tried it out again, and yeah, I can't use ecto upsert because I can't see any obvious way to get it to respect the foreign key constraint for the user.

This is what the code looks like (I would love it for being this concise though):

```diff
   def upsert(user, user_identity_params, config) do
-    params                                   = convert_params(user_identity_params)
-    {uid_provider_params, additional_params} = Map.split(params, ["uid", "provider"])
+    params        = convert_params(user_identity_params)
+    user_identity = Ecto.build_assoc(user, :user_identities)
+    repo_opts     = Config.get(config, :repo_opts, []) ++ [conflict_target: [:uid, :provider], on_conflict: :replace_all_except_primary_key]
+    config        = Config.put(config, :repo_opts, repo_opts)

    user
-    |> get_for_user(uid_provider_params, config)
-    |> case do
-      nil      -> insert_identity(user, params, config)
-      identity -> update_identity(identity, additional_params, config)
-    end
+    |> Ecto.build_assoc(:user_identities)
+    |> user_identity.__struct__.changeset(params)
+    |> Context.do_insert(config)
    |> case do
      {:error, %{errors: [uid_provider: _]} = changeset} ->
        {:error, {:bound_to_different_user, changeset}}

  # ...

  defp convert_param({key, value}) when is_atom(key), do: {Atom.to_string(key), value}
  defp convert_param({key, value}) when is_binary(key), do: {key, value}

-  defp insert_identity(user, user_identity_params, config) do
-    user_identity = Ecto.build_assoc(user, :user_identities)
-
-    user_identity
-    |> user_identity.__struct__.changeset(user_identity_params)
-    |> Context.do_insert(config)
-  end
-
-  defp update_identity(user_identity, additional_params, config) do
-    user_identity
-    |> user_identity.__struct__.changeset(additional_params)
-    |> Context.do_update(config)
-  end
-
-  defp get_for_user(user, %{"uid" => uid, "provider" => provider}, config) do
-    user_identity = Ecto.build_assoc(user, :user_identities).__struct__
-
-    repo(config).get_by(user_identity, user_id: user.id, provider: provider, uid: uid)
-  end
-
```

Would also require a change in [`Pow.Context.do_insert/2`](https://github.com/danschultzer/pow/blob/f5a227833358b5ec2efc39293b0873163a71e9bb/lib/pow/ecto/context.ex#L189-L196) so it can accept the repo opts:

```diff
  @spec do_insert(changeset(), Config.t()) :: {:ok, user()} | {:error, changeset()}
  def do_insert(changeset, config) do
-    opts = repo_opts(config, [:prefix])
+    opts = repo_opts(config, [:prefix, :conflict_target, :on_conflict])

    changeset
    |> Config.repo!(config).insert(opts)
    |> reload_after_write(config)
  end
```

Reply:
I think I don't understand the problem. It'll update the `user_id`, but it'll always update it to the same value, since you're building the association from the user.

Reply:
Alternatively, you could change the `:on_conflict` option to use the `{:replace, fields}` form. You'd just have to build fields yourself. For example, get all the fields from the schema and remove `id` and `user_id`.

Reply:
The `provider`-`uid` may already have been set for another user. The above will override the user id, invalidating provider auth for the previous user (the resource owner is the same, but it's still an issue).

Good idea with the `:on_conflict` option, I'll test it out.

Reply:
Wait, it would probably still update the user identity if I do that. Ideally it should return an error if the user id differs, otherwise upsert the identity.

Reply:
This is the test that fails (user id being overridden instead of returning a `:bound_to_different_user` error):

https://github.com/danschultzer/pow_assent/blob/d1b537304364c8984efba6c008e254d4ebd17fb5/test/pow_assent/ecto/user_identities/context_test.exs#L113-L120

Reply:
Ah, I understand now. I believe this would work:

1. Upsert with `{:replace, all_fields_except_id_and_user_id}`
2. Check if returned user_identity.user_id == user.id
  1. Yes - no error
  2. No - `:bound_to_different_user`

Reply:
I'll think on this a bit

Reply:
Is there a branch with these changes I could play around with?

Reply:
Sure. Just pushed [`use-ecto-upsert`](https://github.com/danschultzer/pow_assent/tree/use-ecto-upsert) with the above changes. It pulls pow from the [`allow-conflict-repo-opts`](https://github.com/danschultzer/pow/tree/allow-conflict-repo-opts) branch that picks up the extra repo opts.

Reply:
I believe the proper way to do this is to include a where clause in the on_conflict option. The way to do that is to pass a query to on_conflict in ecto. I think this would achieve that. Though, I haven't run this code, so not sure how it'd behave.

```
updates = Enum.reject(changeset.changes, fn {key, _} -> key == :id end)
on_conflict = 
  from(ui in UserIdentity, 
    where: ui.user_id != ^(user.id),
    update: [set: updates]
Repo.insert(changeset, conflict_target: [:uid, :provider], on_conflict: on_conflict)
```

Reply:
Moved the discussion to #78. I'll try out your suggestion and see if I can get it working!